SCION-Box Controller & SCION-Box Database

philippmao opened the pull request:
This PR adds the SCION-Box Controller and the SCION-Box Database, 

- new Boxes can use the SCION-Box API to get a gen folder connecting themselves to attachment points in the same ISD

This PR addresses issues #109, #89, #90 and #92.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion-coord/131)
<!-- Reviewable:end -->

philippmao commented:
TODO:  github.com/oschwald/geoip2-golang  needs to be added to the vendor file

FR4NK-W commented:
@philippmao updated the vendor file, added oschwald/geoip2-golang and oschwald/maxminddb-golang which was also required

FR4NK-W commented:
Reviewed 1 of 17 files at r6, 13 of 17 files at r7.
Review status: 14 of 18 files reviewed at latest revision, 4 unresolved discussions.

---

*[main.go, line 56 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0Eul17L51ZGQ7v4CsJ:-L0Eul17L51ZGQ7v4CsK:b-g0nep) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/main.go#L56)):*
> ```Go
> 	for _, ISD := range(isd_loc) {
> 		_, err = models.FindISDbyID(ISD.ISD)
> 		if err != nil {
> ```

This check can be dropped since we handle the orm.ErrNoRows case just next and a genuine error is handled below. 

---

*[conf/development.conf.default, line 83 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0Ev1iWrg8wegraueEf:-L0Ev1iWrg8wegraueEg:b9hpawn) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/conf/development.conf.default#L83)):*
> ```
> 
> # HeartBeat period in minutes
> heartbeat.period = 1
> ```

Do we want to define heartbeat.period in minutes? I think it is easier to define it in seconds and have 60 seconds as default, just in case we want to lower it for any reason in the future.

---

*[templates/simple_box_config_topo.tmpl, line 30 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0EvwLc-T8sd7A4-Qpp:-L0EvwLc-T8sd7A4-Qpq:b-zdrklt) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/templates/simple_box_config_topo.tmpl#L30)):*
> ```
>           "LinkType": "{{.LINK_TYPE}}",
>           "Bandwidth": 1000,
>           "MTU": 1472,
> ```

MTU should probably also be parametrized 

---

*[utility/geolocation/geolocation.go, line 19 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0EwIOQS3s9S9UgMixu:-L0EwIORrdnHypWCXRAO:b-39q2sm) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/utility/geolocation/geolocation.go#L19)):*
> ```Go
> 	}
> 	defer db.Close()
> 	// If you are using strings that may be invalid, check that ip is not nil
> ```

As we already discussed, we could have a default region when the record is nil

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131)*
<!-- Sent from Reviewable.io -->

FR4NK-W commented:
Review status: 14 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[controllers/api/scion_box.go, line 167 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0EzEs8SugURyIlakxD:-L0EzEs8SugURyIlakxE:b-t0wu9f) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L167)):*
> ```Go
> 
> // Receive a Post request with json: {IPAddress: 'string', MacAddress: 'string', OpenPorts: int, StartPort: int}
> func (s *SCIONBoxController) parseRequest(r *http.Request) (bool, string, string, string, int, int, error) {
> ```

Named return values would help here (same names as already used in the return statement)

---

*[controllers/api/scion_box.go, line 316 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0F-OpsIkGUOFGs3IdQ:-L0F-OptVH4hAvMATRyM:bb752ax) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L316)):*
> ```Go
> 		return
> 	}
> 	// Get the Reqeust IP
> ```

Typo

---

*[controllers/api/scion_box.go, line 445 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0F7bscoIJna-cq-3rk:-L0F7bscoIJna-cq-3rl:b-kkcx6v) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L445)):*
> ```Go
> 	}
> 	// Wait to make sure a previous version is not served
> 	time.Sleep(1000 * time.Millisecond)
> ```

Or just time.Second

---

*[controllers/api/scion_box.go, line 468 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0F9MPglD5gxOYlkuSn:-L0F9MPhKfaPlgBrrY_y:b-gvzupi) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L468)):*
> ```Go
> 		return "", err
> 	}
> 	//userIP := net.ParseIP(ip)
> ```

This is not needed since we always use the string representation of IPs, never use the net.IP one (maybe we should, be then we should do it everywhere)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131)*
<!-- Sent from Reviewable.io -->

philippmao commented:
Review status: 8 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[main.go, line 56 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0Eul17L51ZGQ7v4CsJ:-L0FRSag3SvgNmF5KJwN:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/main.go#L56)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

This check can be dropped since we handle the orm.ErrNoRows case just next and a genuine error is handled below. 
</blockquote></details>

Done.

---

*[conf/development.conf.default, line 83 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0Ev1iWrg8wegraueEf:-L0FV4KA6KfmubxZFgWV:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/conf/development.conf.default#L83)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

Do we want to define heartbeat.period in minutes? I think it is easier to define it in seconds and have 60 seconds as default, just in case we want to lower it for any reason in the future.
</blockquote></details>

Done.

---

*[controllers/api/scion_box.go, line 167 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0EzEs8SugURyIlakxD:-L0FW_rK0Md7Vy0c-g_C:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L167)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

Named return values would help here (same names as already used in the return statement)
</blockquote></details>

Done.

---

*[controllers/api/scion_box.go, line 316 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0F-OpsIkGUOFGs3IdQ:-L0FWjU3GemLckk2rhPc:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L316)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

Typo
</blockquote></details>

Done.

---

*[controllers/api/scion_box.go, line 445 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0F7bscoIJna-cq-3rk:-L0FWopbq9HLN1eaYfAD:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L445)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

Or just time.Second
</blockquote></details>

Done.

---

*[controllers/api/scion_box.go, line 468 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0F9MPglD5gxOYlkuSn:-L0FWxSueYEt6oUaXb8t:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L468)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

This is not needed since we always use the string representation of IPs, never use the net.IP one (maybe we should, be then we should do it everywhere)
</blockquote></details>

Done.

---

*[templates/simple_box_config_topo.tmpl, line 30 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0EvwLc-T8sd7A4-Qpp:-L0FXiAKhiJulKshXJst:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/templates/simple_box_config_topo.tmpl#L30)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

MTU should probably also be parametrized 
</blockquote></details>

Done.

---

*[utility/geolocation/geolocation.go, line 19 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0EwIOQS3s9S9UgMixu:-L0FZXypSU1eLjWgP3ot:b-xjls4p) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/utility/geolocation/geolocation.go#L19)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

As we already discussed, we could have a default region when the record is nil
</blockquote></details>

added a default entry to the ISD-location database. {ISD:1, Continent: "", Country:""}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131)*
<!-- Sent from Reviewable.io -->

FR4NK-W commented:
Reviewed 2 of 17 files at r7.
Review status: 10 of 18 files reviewed at latest revision, 10 unresolved discussions.

---

*[controllers/api/scion_box.go, line 496 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0FZf7cMp3EbnDFSxAK:-L0FZf7dGQHUFCK668NP:bm1kt4n) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L496)):*
> func (s *SCIONBoxController) findLowestBRId(slas *models.SCIONLabAS) int {
> 	var newIdFound = false
> 	var ID = 1
> 	var idFound bool
> 	cns, _ := slas.GetConnectionInfo()
> 	for !newIdFound {
> 		idFound = false
> 		for _, cn := range cns {
> 			if cn.Status != models.REMOVED {
> 				if cn.BRID == ID {
> 					idFound = true
> 					break
> 				}
> 			}
> 		}
> 		if idFound {
> 			ID++
> 		} else {
> 			newIdFound = true
> 		}
> 	}
> 	return ID
> }


This can be simplified quite a bit to something like this:

>	var ID = 1
>	cns, _ := slas.GetConnectionInfo()
>	for {
>		idFound := true
>		for _, cn := range cns {
>			if cn.Status != models.REMOVED && cn.BRID == ID {
>				idFound = false
>				ID++
>				break
>			}
>		}
>		if idFound {
>			return ID
>		}
	}

---

*[utility/topologyAlgorithm/topology.go, line 153 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0FaJZsyu4rCCZDXoUz:-L0FaJZsyu4rCCZDXoV-:b-44bm5z) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/utility/topologyAlgorithm/topology.go#L153)):*
> ```Go
> 
> // Get the number of neighbors from the database
> // If an error occurs return 9999
> ```

Wouldn't -1 in case of error make more sense? The number of neighbors will not grow up to 9999, but I also don't think that we can interpret no entry in the DB as a good sign, which is what happens here when used in comparisons like 'bestNb.getDegree() > nb.getDegree()'

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131)*
<!-- Sent from Reviewable.io -->

philippmao commented:
Review status: 10 of 18 files reviewed at latest revision, 10 unresolved discussions.

---

*[controllers/api/scion_box.go, line 496 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0FZf7cMp3EbnDFSxAK:-L0FkgdOo7hxTSNkIIBG:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/controllers/api/scion_box.go#L496)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

> func (s *SCIONBoxController) findLowestBRId(slas *models.SCIONLabAS) int {
> 	var newIdFound = false
> 	var ID = 1
> 	var idFound bool
> 	cns, _ := slas.GetConnectionInfo()
> 	for !newIdFound {
> 		idFound = false
> 		for _, cn := range cns {
> 			if cn.Status != models.REMOVED {
> 				if cn.BRID == ID {
> 					idFound = true
> 					break
> 				}
> 			}
> 		}
> 		if idFound {
> 			ID++
> 		} else {
> 			newIdFound = true
> 		}
> 	}
> 	return ID
> }


This can be simplified quite a bit to something like this:

>	var ID = 1
>	cns, _ := slas.GetConnectionInfo()
>	for {
>		idFound := true
>		for _, cn := range cns {
>			if cn.Status != models.REMOVED && cn.BRID == ID {
>				idFound = false
>				ID++
>				break
>			}
>		}
>		if idFound {
>			return ID
>		}
	}


</blockquote></details>

Done.

---

*[utility/topologyAlgorithm/topology.go, line 153 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-L0FaJZsyu4rCCZDXoUz:-L0FkhSEQjdLgRJBcD1h:b57rrpz) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/2e28099f22df0813424141e38bc9e245804d8720/utility/topologyAlgorithm/topology.go#L153)):*
<details><summary><i>Previously, FR4NK-W wrote…</i></summary><blockquote>

Wouldn't -1 in case of error make more sense? The number of neighbors will not grow up to 9999, but I also don't think that we can interpret no entry in the DB as a good sign, which is what happens here when used in comparisons like 'bestNb.getDegree() > nb.getDegree()'
</blockquote></details>

I agree that setting it to -1 would make it look cleaner but it then I need another check in between which I wanted to avoid. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131)*
<!-- Sent from Reviewable.io -->

FR4NK-W commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 17 files at r7, 7 of 7 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131#-:-L1xxQiekeTu0nhhVgEa:bnfp4nl)*
<!-- Sent from Reviewable.io -->

FR4NK-W commented:
Reviewed 2 of 2 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/131)*
<!-- Sent from Reviewable.io -->

hausheer commented:
As discussed with @mlegner, @FR4NK-W, and @philippmao the conflicting changes with PR #136 should be addressed.
